source/robot_model/src/Model.cpp

```diff
@@ -192,65 +192,78 @@ StateRepresentation::CartesianPose Model::forward_geometry(const StateRepresenta
-    // 'Infeasibility-free inverse kinematics method'.
-
-    const int JOINT_ID = this->robot_model_.njoints - 2;
-
-    // TODO if the actual_joint_state is not provided initialize the position of the robot in a radom conformation
-    // TODO take as input the parameters eps and IT_MAX
-    // TODO return 'success'
-
-    Eigen::VectorXd q = actual_joint_state.get_positions(); //pinocchio::neutral(model);
-    const double eps = 1e-4;
-    const int IT_MAX = 1000;
-
-    std::vector<unsigned int> frame_ids{(unsigned int)JOINT_ID};
-
-    StateRepresentation::CartesianPose actual_cartesian_state;
-    double error;
-
-    bool success = false;
-    Eigen::VectorXd v(this->robot_model_.nv);
-    for (int i = 0;; i++)
-    {
-
-      actual_cartesian_state = this->forward_geometry(actual_joint_state, frame_ids).front();
-      error = desired_cartesian_state.dist(actual_cartesian_state, StateRepresentation::CartesianStateVariable::POSE);
-
-      if (error < eps)
-      {
-        success = true;
-        break;
-      }
-      if (i >= IT_MAX)
-      {
-        success = false;
-        break;
-      }
-      v = this->inverse_kinematic(actual_joint_state, desired_cartesian_state).get_velocities();
-      q = actual_joint_state.get_positions();
-      actual_joint_state.set_velocities(v);
-      StateRepresentation::CartesianTwist x_dot = this->forward_kinematic(actual_joint_state);
-
-      StateRepresentation::CartesianTwist zerovec;
-      zerovec.Zero("zero");
-      double x_dot_norm = x_dot.dist(zerovec, StateRepresentation::CartesianStateVariable::POSE);
-
-      // dt=||x_d - x||/||x'||
-      double dt = error / x_dot_norm;
-
-      q = pinocchio::integrate(this->robot_model_, q, v * dt);
-      actual_joint_state.set_positions(q);
-    }
-    return StateRepresentation::JointPositions(actual_joint_state);
+StateRepresentation::JointPositions Model::inverse_geometry(const StateRepresentation::CartesianState& desired_cartesian_state,
+                                                            const StateRepresentation::JointState& current_joint_state,
+                                                            std::string frame_name, const double& tolerance, const int& max_number_of_iteration) {
+
+  if (frame_name.empty()) {
+    // get last frame if none specified
+    frame_name = this->robot_model_.frames.back().name;
+  }else {
+    // throw error if specified frame does not exist
+    if (!this->robot_model_.existFrame(frame_name)) { throw (Exceptions::FrameNotFoundException(frame_name)); }
   }
 
+  Eigen::VectorXd q = current_joint_state.get_positions();
+
+  const int frame_id = this->robot_model_.getFrameId(frame_name);
+  
+  const pinocchio::SE3 oMdes(desired_cartesian_state.get_orientation().toRotationMatrix(), desired_cartesian_state.get_position());
+
+  double dt;
+  // damping added to the diagonal of J*Jt in order to avoid the singularity
+  const double damp = 1e-6;
```

Comment:
This should not be hardcoded. It should be either a parameter of the function or a parameter of the class.

Reply:
Yes exactly and I wanted to discuss with you how do we do... can I create a structure which contains all the parameters and we enter the structure as parameter of the function?

Reply:
For the sake of simplicity in the first version, my suggestion was to keep some of these solver parameters fixed for now - I don't think they make sense as Model member properties, and having many input / default arguments is not necessarily good function design. I also think it may be the case that something like this damping rarely needs to be set by the user.

A solution is indeed to add a custom IK parameter structure to contain all these fields. But, I think this should wait for the next PR. My main interest is that this PR does not inflate in scope, and to get the minimum viable version approved (correct formatting and working tests). I'm worried that adding extra patterns and logic will delay this further.

To sum up: some simple and consistent behaviour that is finished is better than parametrisable and complex behaviour that is unfinished.